### PR TITLE
rewrote singleton objects tour

### PR DESCRIPTION
**Pull request by travissarles**

### CODE REVIEW

File: tutorials/tour/_posts/2017-02-13-singleton-objects.md [hunk 1]
```diff
@@ -10,65 +10,113 @@ num: 13
 
 next-page: xml-processing
 previous-page: pattern-matching
+assumed-knowledge: classes, methods, private-methods, packages
 ---
 
-Methods and values that aren't associated with individual instances of a [class](classes.html) belong in *singleton objects*, denoted by using the keyword `object` instead of `class`.
+Singleton objects are like classes in that they contain methods and values, but there are a few key differences.
+* There is only one instance of a singleton object so their methods and values aren't associated with individual instances of a [class](classes.html).
+* Objects have no constructors so they cannot be instantiated
```
Comment (som-snytt) [inline tutorials/tour/_posts/2017-02-13-singleton-objects.md]: This language has to be wrong and misleading. The spec says it nicely: "a single object of a new class." I might have preferred "a single instance of a new class." Occasionally, you have to be aware of when object init happens, including common use cases with circular definitions.

Comment (travis032654) [inline tutorials/tour/_posts/2017-02-13-singleton-objects.md]: @som-snytt Looking back on it, I agree that it's a bit misleading. I like "a single instance of a new class" except that the word "class" could be confusing for people who assume that classes can be instantiated. I'll try to explain that. Thanks for the feedback.

File: tutorials/tour/_posts/2017-02-13-singleton-objects.md [hunk 2]
```diff
@@ -10,65 +10,111 @@ num: 13
...
 previous-page: pattern-matching
+assumed-knowledge: classes, methods, private-methods, packages
 ---
 
-Methods and values that aren't associated with individual instances of a [class](classes.html) belong in *singleton objects*, denoted by using the keyword `object` instead of `class`.
+A singleton object is an instance of a new class. There is exactly one instance of each singleton object. They do not have constructors so they cannot be instantiated.
 
+# Defining a singleton object
+Like with classes, the simplest form of an object is the keyword `object` and an identifier:
+```tut
```
Comment (Ichoran) [inline tutorials/tour/_posts/2017-02-13-singleton-objects.md]: Why `tut`?

File: tutorials/tour/_posts/2017-02-13-singleton-objects.md [hunk 3]
```diff
@@ -10,65 +10,111 @@ num: 13
...
+
+val event = new Event("Dotty Days", LocalDateTime.of(2020, 2, 2, 9, 0, 0))
+event.isOnTheWeekend  // true
 ```
-class X {
-  import X._
 
-  def blah = foo
-}
+The `class Event` imports `_isOnTheWeekend` from the `object Event`. To make a member inaccessible from outside, use `private[this]`.
```
Comment (Ichoran) [inline tutorials/tour/_posts/2017-02-13-singleton-objects.md]: To make it inaccessible _from its companion class_ you mean.  It is inaccessible everywhere else with just `private`

File: tutorials/tour/_posts/2017-02-13-singleton-objects.md [hunk 4]
```diff
@@ -10,65 +10,111 @@ num: 13
...
+
+Note: If a class or object has a companion, both must be defined in the same file. To define them in the REPL, you must enter `:paste` and then paste in the class and companion object code.
+
+## Sharing global state
```
Comment (Ichoran) [inline tutorials/tour/_posts/2017-02-13-singleton-objects.md]: This is such a bad practice I'm not sure it's a good idea to give it as an example.  If yes, I'd change the tone and make it "Beware of global state", and end with a thread-safe example.

File: tutorials/tour/_posts/2017-02-13-singleton-objects.md (hunk 2 above)
Comment (SethTisue) [inline tutorials/tour/_posts/2017-02-13-singleton-objects.md]: we should be marking code blocks `tut` whenever possible, so PR validation checks that the code compiles. see https://github.com/scala/scala.github.com/issues/677

File: tutorials/tour/_posts/2017-02-13-singleton-objects.md (hunk 3 above)
Comment (travis032654) [inline tutorials/tour/_posts/2017-02-13-singleton-objects.md]: Right, I'll change the wording.

File: tutorials/tour/_posts/2017-02-13-singleton-objects.md [hunk 5]
```diff
@@ -10,65 +10,85 @@ num: 13
...
 next-page: xml-processing
 previous-page: pattern-matching
+assumed-knowledge: classes, methods, private-methods, packages
 ---
 
-Methods and values that aren't associated with individual instances of a [class](classes.html) belong in *singleton objects*, denoted by using the keyword `object` instead of `class`.
+A singleton object is an instance of a new class. There is exactly one instance of each singleton object. They do not have constructors so they cannot be instantiated.
 
+# Defining a singleton object
+Like with classes, the simplest form of an object is the keyword `object` and an identifier:
```
Comment (julienrf) [inline tutorials/tour/_posts/2017-02-13-singleton-objects.md]: I don’t really understand the connection with classes, so I wouldn’ start with “like with classes”.

File: tutorials/tour/_posts/2017-02-13-singleton-objects.md [hunk 6]
```diff
@@ -10,65 +10,85 @@ num: 13
...
+A common pattern to simplify calls to the companion object's members from the companion class is to import everything:
 ```
-class X {
-  import X._
+import java.time.LocalDateTime;
+import java.time.DayOfWeek
 
-  def blah = foo
+object Event {
+  private def _isOnTheWeekend(date: LocalDateTime) = Array(DayOfWeek.SATURDAY, DayOfWeek.SUNDAY).contains(date.getDayOfWeek)
```
Comment (julienrf) [inline tutorials/tour/_posts/2017-02-13-singleton-objects.md]: I’m not sure to see any advantage of putting this as a private def in the companion object. I think that maybe more compelling use cases for companion object’s methods are constants definition, default instances and alternative constructors.

**Comment by travis032654** [comment]
@heathermiller ready to merge

**Comment by SethTisue** [comment]
Both the old and new texts assume that `object` appears at the top level, in which case it is truly a singleton classloader-wide. But you can also nest `object` within other constructs and then there's more than one — it's only a singleton _in that context_. 

Since the old text also had this problem, I think this PR could be merged anyway, but Travis, perhaps you're motivated to add some wording that addresses this? If not, perhaps we (or someone) can come back to it later.

Otherwise LGTM for merge.

**Comment by som-snytt** [comment]
Rebased at https://github.com/scala/docs.scala-lang/pull/978

**Pull request closed without merging**